base/compiler/typeinfer.jl

```diff
@@ -372,23 +371,30 @@ function transform_result_for_cache(interp::AbstractInterpreter,
 end
 
 function cache_result!(interp::AbstractInterpreter, result::InferenceResult)
-    valid_worlds = result.valid_worlds
-    if last(valid_worlds) == get_world_counter()
+    if last(result.valid_worlds) == get_world_counter()
         # if we've successfully recorded all of the backedges in the global reverse-cache,
         # we can now widen our applicability in the global cache too
-        valid_worlds = WorldRange(first(valid_worlds), typemax(UInt))
+        result.valid_worlds = WorldRange(first(result.valid_worlds), typemax(UInt))
     end
     # check if the existing linfo metadata is also sufficient to describe the current inference result
     # to decide if it is worth caching this
     mi = result.linfo
     already_inferred = already_inferred_quick_test(interp, mi)
-    if !already_inferred && haskey(WorldView(code_cache(interp), valid_worlds), mi)
-        already_inferred = true
+    cache = WorldView(code_cache(interp), result.valid_worlds)
+    if !already_inferred && haskey(cache, mi)
+        ci = cache[mi]
+        # Even if we already have a CI for this, it's possible that the new CI has more
+        # information (E.g. because the source was limited before, but is no longer - this
+        # happens during bootstrap). In that case, allow the result to be recached.
+        if result.src === nothing || (ci.inferred !== nothing || ci.invoke != C_NULL)
```

Comment:
Why these particular conditions? Is this what `already_inferred_quick_test` is already supposed to be checking?

Reply:
`already_inferred_quick_test` handles the case where you're trying to cache the same MI twice in the same call to typeinf_ext (it just checks `mi.inInference`). However, this case is different. Here we check whether we need to upgrade the CodeInstance that's already in the cache to add source to it. This can happen because the last time we inferred it, we couldn't save the source because one of our (unused) body statements had a limited rettype. The condition is essentially `ci_has_abi`. I should probably replace it.

Reply:
I don't think the cache override is unique to the bootstrapping process. You can create the same situation at any time by first inferring something that will poison the the code and then it gets re-inferred when the execution actually reaches.

Reply:
The pair of conditions though `(ci.inferred !== nothing || ci.invoke != C_NULL)` seems to be not something that has a well-defined meaning in a multithread program (e.g. it introduces a data-race here).